webserver/src/main.rs

```diff
@@ -0,0 +1,417 @@
+use std::ops::Range;
+use tide::http::Mime;
+use tide::{Request, Response, StatusCode};
+use tide_jsx::*;
+use time::OffsetDateTime;
+
+/// Simple program to greet a person
+#[derive(Parser, Debug)]
+#[command(author, version, about, long_about = None)]
+struct Args {
+   #[arg(long, default_value = "0.0.0.0:38618")]
+   bind_address: String,
+
+   /// Number of times to greet
+   #[arg(short, long)]
+   database_address: String,
+}
+
+#[async_std::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    femme::start();
+    let args = Args::parse();
+    let mut app = tide::with_state(RusqliteReader::new(&args.database_address)?);
+    app.with(tide::log::LogMiddleware::new());
+
+    app.with(tide::utils::After(|mut res: Response| async {
+        if let Some(err) = res.error() {
+            let msg = format!("Error: {:?}", err);
+            res.set_status(err.status());
+            res.set_body(msg);
+        }
+        Ok(res)
+    }));
+
+    app.at("/dyn/games").get(games);
+    app.at("/dyn/pgn/:rowid").get(pgn);
+    app.at("/dyn/stats").get(|r| stats(r, None));
+    app.at("/dyn/stats/:duration").get(stats_with_duration);
+    app.listen(args.bind_address).await?;
+    Ok(())
+}
+
+#[derive(Copy, Clone)]
+struct RowId {
+    id: i64,
+}
+
+trait DatabaseReader {
+    // TODO: An API with bounded response size
+    fn finished_games(
+        &self,
+        game_start_time_range: Range<OffsetDateTime>,
+    ) -> Result<Vec<(RowId, GameResultRow)>, anyhow::Error>;
+
+    fn pgn(&self, rowid: RowId) -> Result<String, anyhow::Error>;
+}
+
+#[derive(Clone)]
+struct RusqliteReader {
+    // TODO: move to sqlx or similar to take advantage of concurrency.
+    // This blocks concurrent requests.
+    conn: std::sync::Arc<std::sync::Mutex<Connection>>,
+}
+
+impl RusqliteReader {
+    pub fn new(db_address: &str) -> Result<Self, anyhow::Error> {
+        let conn = rusqlite::Connection::open(db_address)?;
+        Ok(Self {
+            conn: std::sync::Arc::new(std::sync::Mutex::new(conn)),
+        })
+    }
+}
+
+impl DatabaseReader for RusqliteReader {
+    fn finished_games(
+        &self,
+        game_start_time_range: Range<OffsetDateTime>,
+    ) -> Result<Vec<(RowId, GameResultRow)>, anyhow::Error> {
+        let conn = self.conn.lock().unwrap();
+        let mut stmt = conn.prepare(
+            "SELECT *, ROWID from finished_games
+             WHERE game_start_time >= ?1 AND game_start_time < ?2
+             ORDER BY game_start_time DESC",
+        )?;
+        let (oks, errs): (Vec<_>, _) = stmt
+            .query_map(
+                (
+                    game_start_time_range.start.unix_timestamp(),
+                    game_start_time_range.end.unix_timestamp(),
+                ),
+                |row| {
+                    Ok((
+                        RowId { id: row.get(10)? },
+                        GameResultRow {
+                            git_version: row.get(0)?,
```

Comment:
I'm a bit scared of unpacking the entire row based on column indices. Maybe replace “SELECT *” with column list?

Reply:
Right. I've remove pgns for now, so that it's not querying the currently unused column that's also the biggest.

Also, get() supports &str as the column index type, so that I don't need to rely on ordering here.